File: dsa-problems/LinkedLists/solutions/javascript/LNK-002-campus-badge-search.js
const readline = require("readline");

class ListNode {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

function findFirstIndex(head, target) {
  let current = head;
  let index = 0;
  while (current !== null) {
    if (current.val === target) {
      return index;
    }
    current = current.next;
    index++;
  }
  return -1;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/).filter(x => x)));
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  
  const dummy = new ListNode(0);
  let cur = dummy;
  for (let i = 0; i < n; i++) {
    cur.next = new ListNode(parseInt(data[idx++], 10));
    cur = cur.next;
  }
  
  if (idx < data.length) {
      const target = parseInt(data[idx++], 10);
      console.log(findFirstIndex(dummy.next, target));
  }
});

File: dsa-problems/LinkedLists/solutions/LNK-002-campus-badge-search.py
import sys

class ListNode:
    def __init__(self, val=0):
        self.val = val
        self.next = None

def find_first_index(head: ListNode, target: int) -> int:
    current = head
    index = 0
    while current:
        if current.val == target:
            return index
        current = current.next
        index += 1
    return -1

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return

    iterator = iter(data)
    try:
        n = int(next(iterator))
        dummy = ListNode()
        cur = dummy
        for _ in range(n):
            val = int(next(iterator))
            cur.next = ListNode(val)
            cur = cur.next
        
        target = int(next(iterator))
        print(find_first_index(dummy.next, target))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()
